dagshub/common/api/repo.py

```diff
@@ -388,6 +388,30 @@ def download(
         download_files(file_tuples, skip_if_exists=not redownload)
         log_message(f"Downloaded {len(files)} file(s) to {local_path.resolve()}")
 
+    def import_metadata_from_file(self, datasource_name: str, file_path: Union[str, PathLike], path_column: str):
+        files = {
+            "file": (
+                Path(file_path).name,
+                open(file_path, "rb"),
+                mimetypes.guess_type(file_path)[0] or "application/octet-stream",
+            ),
+        }
+
+        data = {
+            "datasource_name": datasource_name,
+            "path_column": path_column,
+        }
+
+        res = self._http_request("POST", self.repo_import_metadata_from_file_url(), data=data, files=files)
+        if res.status_code == 404:
+            raise PathNotFoundError(f"Datasource {datasource_name} not found")
+        elif res.status_code >= 400:
+            error_msg = f"Got status code {res.status_code} when importing metadata to datasource {datasource_name}"
+            logger.error(error_msg)
+            logger.debug(res.content)
+            raise RuntimeError(error_msg)
+        return res.content
```

Comment:
How would you do that?

Do you mean call existing `RepoAPI.import_metadata_from_file` from `DatasourceState` or implement the request itself inside `DatasourceState`?

Reply:
Second one, just cut-paste the function into there.
I know that the name is `DatasourceState`, but I kind of made it to be a place where you would have most of the lower-level interactions with the data engine backend. Meanwhile the datasource is what the user interacts with.

Reply:
Copy-paste won't work out of the box because `DatasourceState` doesn't have `self._http_request`. It uses the REST API via `self.repoApi` and GQL API via `self.client`.

How would you tackle that?
Adding or re-using`self._http_request` seems just wrong because we have `self.repoApi` for that.
Creating a wrapper function in `DatasourceState` seems redundant overhead, but at least it would keep using the convention that REST API calls made via `self.repoApi`.

WDYT?

Reply:
`DatasourceState` still has `repoAPI` in it, so it can call `state.repoApi._http_request`.
It's not really a good practice to use underscored functions, but it's not like they are actually private, you can still call them.

My problem with having this function in RepoAPI is that RepoAPI might be useful for a user, so we're trying to keep the functions there useful for a regular user and have them make sense. This function is too tightly bound to Data Engine to really be put in RepoAPI imo, since it expects to have a datasource specified in the first place.

Keep in mind that this function will basically show up here, after you write a docstring for it: https://dagshub.com/docs/client/reference/repo_api.html and it will also show up in autocompletion.

IMO putting this in `DatasourceState` away from public functions, and using a private `repoAPI._http_request` is the best compromise